lib/graph-document.js

```diff
@@ -21,92 +22,100 @@ function node (state, createEdge) {
-      if (err) return self.emit(err)
-      createEdge('list', Buffer.from(list.join(',')))
-    })
+  var fonts = extractFonts(state.assets)
+  var list = ssr.routes
 
-    function documentifyRoute (route, done) {
-      renderRoute(route, function (err, buf) {
-        if (err) return done(err)
+  mapLimit(list, WRITE_CONCURRENCY, documentifyRoute, function (err) {
+    if (err) return self.emit(err)
+    createEdge('list', Buffer.from(list.join(',')))
+  })
 
-        var name = route
-        if (name === '/') name = 'index'
-        name = name + '.html'
-        createEdge(name, buf)
-        done()
+  function documentifyRoute (route, done) {
+    waterfall([
+      renderApp,
+      documentifyApp,
+      pushEdge
+    ], done)
+
+    function renderApp (done) {
+      ssr.render(route, function (err, content) {
+        if (err) {
+          self.emit('ssr', {success: false, error: err})
+          return done(err)
+        }
+        done(null, content)
       })
     }
 
-    function renderRoute (route, done) {
-      var res
-
-      try {
-        res = render(route) // TODO: allow passing in state as second arg
-      } catch (err) {
-        self.emit('ssr', {success: false, error: err})
-        return done(err)
-      }
-
-      var language = res.language
-      var title = res.title
-      var body = res.body
-
-      var hasDynamicScripts = state.scripts.list.buffer.length > 0
-
-      var html = head(body, language)
-      var d = documentify(entry, html)
-      var header = [
-        viewportTag(),
-        scriptTag({ hash: state.scripts.bundle.hash }),
-        hasDynamicScripts && dynamicScriptsTag({
-          bundleNames: String(state.scripts.list.buffer).split(','),
-          scripts: state.scripts
-        }),
-        preloadTag(),
-        loadFontsTag({ fonts: fonts }),
-        manifestTag(),
-        descriptionTag({ description: String(state.manifest.description.buffer) }),
-        themeColorTag({ color: String(state.manifest.color.buffer) }),
-        titleTag({ title: title })
-      ].filter(Boolean)
-      // TODO: twitter
-      // TODO: facebook
-      // TODO: apple touch icons
-      // TODO: favicons
-
-      if (state.metadata.reload) {
-        header.push(reloadTag({ bundle: state.reload.bundle.buffer }))
-      }
-
-      d.transform(addToHead, header.join(''))
-
-      if (state.styles.bundle.buffer.length) {
-        d.transform(criticalTransform, { css: state.styles.bundle.buffer })
-      }
-
-      d.transform(addToHead, styleTag({ hash: state.styles.bundle.hash }))
-
-      function complete (buf) { done(null, buf) }
-
-      pump(d.bundle(), concat({ encoding: 'buffer' }, complete), function (err) {
-        if (err) return done(explain(err, 'Error in documentify while operating on ' + route))
-      })
+    function pushEdge (buf, done) {
+      var name = route
+      if (name === '/') name = 'index'
+      name = name + '.html'
+      createEdge(name, buf)
```

Comment:
we're getting as far as having the correct buffer shop up here by the way. But then when calling the `./http.js` methods it just sends down an empty body. Looks like it never calls back or something; not  sure what's up with that.

Reply:
yep you were right, it works if i pass the buffer back to the `ssr.renderRoute` call in the `done` callback one line below here

Reply:
nice!